### PR TITLE
Change type of ingest doc meta-data field 'TIMESTAMP' to `Date`

### PR DESCRIPTION
**Pull request by danielmitterdorfer**
With this commit we change the data type of the 'TIMESTAMP'
meta-data field from a formatted date string to a plain
`java.util.Date` instance. The main reason for this change is
that our benchmarks have indicated that this contributes
significantly to the time spent in the ingest pipeline.

The overhead in terms of indexing throughput of the ingest
pipeline is about 15% and breaks down roughly as follows:

* 5% overhead caused by the conversion from `XContent` -> `Map`
* 5% overhead caused by the timestamp formatting (eliminated by this PR)
* 5% overhead caused by the conversion `Map` -> `XContent`

Relates #22074

### CODE REVIEW

**Comment by danielmitterdorfer** [comment]
@martijnvg Can you please review this one? As this is - strictly speaking - a breaking change I'm not sure whether we should backport it 5.x.

File: core/src/main/java/org/elasticsearch/ingest/IngestDocument.java [hunk 1]
```diff
@@ -68,9 +64,7 @@ public IngestDocument(String index, String type, String id, String routing, Stri
         }
 
         this.ingestMetadata = new HashMap<>();
-        DateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZZ", Locale.ROOT);
-        df.setTimeZone(TimeZone.getTimeZone("UTC"));
-        this.ingestMetadata.put(TIMESTAMP, df.format(new Date()));
+        this.ingestMetadata.put(TIMESTAMP, new Date());
```
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/ingest/IngestDocument.java]: Strictly speaking this isn't a breaking change.

In `PipelineExecutionService#innerExecute(...)` when we serialize the map, the `Date` instance will be serialized by `XContentBuilder#DEFAULT_DATE_PRINTER` formatter, which uses this format `yyyy-MM-dd'T'HH:mm:ss.SSSZZ`, so things remain to be the same.

File: docs/reference/migration/migrate_6_0/ingest.asciidoc [hunk 2]
```diff
@@ -0,0 +1,6 @@
+[[breaking_60_ingest_changes]]
+=== Ingest changes
+
+==== Timestamp meta-data field type has changed
+
+The type of the "timestamp" meta-data field has changed from `java.lang.String` to `java.util.Date`.
```
Comment (martijnvg) [inline docs/reference/migration/migrate_6_0/ingest.asciidoc]: maybe mark this is a java breaking change? I think only authors of custom ingest processors will notice this change.

**Review by martijnvg — APPROVED** [review]
Thanks @danielmitterdorfer!

LGTM. I think this can be backported to 5.x, as this is just a java breaking change.

File: docs/reference/migration/migrate_6_0/ingest.asciidoc (hunk 2 above)
Comment (danielmitterdorfer) [inline docs/reference/migration/migrate_6_0/ingest.asciidoc]: Thanks, I marked it as `breaking-java` now.

**Comment by danielmitterdorfer** [comment]
Thanks for your review @martijnvg! I'll backport the change then to 5.x.

**Pull request merged**